internal/provider/resource_cloudinit_config.go

```diff
@@ -0,0 +1,170 @@
+			},
+			"base64_encode": schema.BoolAttribute{
+				PlanModifiers: []planmodifier.Bool{
+					boolplanmodifier.RequiresReplace(),
+					attribute_plan_modifier_bool.DefaultValue(types.BoolValue(true)),
+				},
+				Optional:            true,
+				Computed:            true,
+				MarkdownDescription: "Specify whether or not to base64 encode the `rendered` output. Defaults to `true`, and cannot be disabled if gzip is `true`.",
+			},
+			"boundary": schema.StringAttribute{
+				Validators: []validator.String{
+					stringvalidator.LengthAtLeast(1),
+				},
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+					attribute_plan_modifier_string.DefaultValue(types.StringValue("MIMEBOUNDARY")),
+				},
+				Optional:            true,
+				Computed:            true,
+				MarkdownDescription: "Specify the Writer's default boundary separator. Defaults to `MIMEBOUNDARY`.",
+			},
+			"rendered": schema.StringAttribute{
+				Computed:            true,
+				MarkdownDescription: "The final rendered multi-part cloud-init config.",
+			},
+			"id": schema.StringAttribute{
+				Computed:            true,
+				MarkdownDescription: "[CRC-32](https://pkg.go.dev/hash/crc32) checksum of `rendered` cloud-init config.",
+			},
+		},
+		DeprecationMessage: `This resource is deprecated, please use the cloudinit_config data source instead.`,
+		MarkdownDescription: "Renders a [multi-part MIME configuration](https://cloudinit.readthedocs.io/en/latest/explanation/format.html#mime-multi-part-archive) " +
+			"for use with [cloud-init](https://cloudinit.readthedocs.io/en/latest/).\n\n" +
+			"Cloud-init is a commonly-used startup configuration utility for cloud compute instances. It accepts configuration via provider-specific " +
+			"user data mechanisms, such as `user_data` for Amazon EC2 instances. Multi-part MIME is one of the data formats it accepts. For more information, " +
+			"see [User-Data Formats](https://cloudinit.readthedocs.io/en/latest/explanation/format.html) in the cloud-init manual.\n\n" +
+			"This is not a generalized utility for producing multi-part MIME messages. It's feature set is specialized for cloud-init multi-part MIME messages.",
+	}
+}
+
+func (r *configResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var newState configModel
+
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &newState)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(updateConfigModel(ctx, &newState)...)
+	resp.Diagnostics.Append(resp.State.Set(ctx, newState)...)
+}
+
+func (r *configResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var newState configModel
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &newState)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(updateConfigModel(ctx, &newState)...)
```

Comment:
Is it possible/necessary for the model to be updated during refresh? Since content is a requires replacement string and not say, a path string to a file that can be read with potentially new contents, I'm not sure if anything needs to be done other than roundtripping the state during refresh.

Reply:
Interesting, you're 100% right that it doesn't need to be updated during refresh. I was originally trying to preserve the [datasource shim](https://github.com/hashicorp/terraform-plugin-sdk/blob/v2.24.1/helper/schema/data_source_resource_shim.go#L21) functionality that was originally, but I might have missed the mark on it.

Can I just leave the Read function blank to pass state through?

Reply:
I believe the logic is optional, but it might be good to leave in for clarity in case the resource is ever "fully" created.

https://github.com/hashicorp/terraform-plugin-framework/blob/3c4ae88946c8a2ce1227d5b37c013997f0a46fa6/internal/fwserver/server_readresource.go#L71-L76